lib/gambit/list/list.scm

```diff
@@ -607,22 +607,34 @@
                 (macro-check-list x 2 (assoc obj lst c)
                   #f))))))))
 
-(define-prim (##remq elem lst)
-
-  (include "~~lib/gambit/prim/prim#.scm") ;; map fx+ to ##fx+, etc
-
-  (define (rem elem lst)
-    (if (pair? lst)
-        (let* ((lst2 (rem elem (cdr lst)))
-               (x (car lst)))
-          (if (eq? x elem)
-              lst2
-              (if (eq? lst2 (cdr lst))
-                  lst
-                  (cons x lst2))))
-        '()))
-
-  (rem elem lst))
+(define-macro (##macro-filter prim-call list var test-expr)
+  `(let ()
+     (define (filter-tail list)
+       (if (pair? list)
+           (let* ((,var (car list))
+                  (old-tail (cdr list))
+                  (new-tail (filter-tail old-tail)))
+             (if (and (or (pair? new-tail) (null? new-tail))
+                      ,test-expr)
+                 (if (eq? old-tail new-tail)
+                     list
+                     (cons x new-tail))
+                 new-tail))
+           list))
+     (let ((new-list (filter-tail ,list)))
+       (macro-if-checks (if (or (pair? new-list) (null? new-list))
+                            new-list
+                            (macro-fail-check-list 2 ,prim-call))
+                        new-list))))
+
+(define-prim&proc (filter (pred procedure) (list object))
```

Comment:
I think it is wrong to indicate that `list` is just an `object`.  It would be better to indicate `proper-list` and to have a `macro-check-proper-list` macro that doesn't do any type checking since it will be done later.  At least that way the procedure header contains a type indication that could be used by other tools (like a document generator).  For now the `macro-check-proper-list` macro could be defined locally in `list.scm` like this:

```
(define-syntax macro-check-proper-list
  (lambda (src)
    (##deconstruct-call
     src
     5
     (lambda (arg-src pos-src call-src body-src)
       body-src))))
```

Identical definitions should exist for `dotted-list` and `circular-list`.  To avoid confusion the current macros `macro-check-list` and `macro-check-pair-list` (which do a type check for `null?` and `pair?` respectively but with an exception message saying a "LIST" was expected) should be renamed to `macro-check-empty-proper-list` and `macro-check-nonempty-proper-list`. That can wait for another PR.

Reply:
I tried to add this macro and give `proper-list` as the type in `define-prim&proc` but got segfaults and didn't manage to figure out why it's not working. Should we also add `proper-list` to `##type-exception-names` in `_repl.scm`?

Reply:
@feeley Adding `proper-list` seems more difficult than the rest of this PR. Can you merge this and add the `proper-list`/`dotted-list`/`circular-list` checks? If you can add each to one procedure, I can fill in the rest by looking at how those examples do it.